pkg/controller/controller_binding.go

```diff
@@ -581,14 +581,6 @@ func (c *controller) reconcileServiceBinding(binding *v1beta1.ServiceBinding) er
 
 		// The Bind request has returned successfully from the Broker. Continue
 		// with the success case of creating the ServiceBinding.
-
-		// Save off the external properties here even if the subsequent
-		// credentials injection fails. The Broker has already processed the
-		// request, so this is what the Broker knows about the state of the
-		// binding.
-		toUpdate.Status.ExternalProperties = toUpdate.Status.InProgressProperties
```

Comment:
I was expecting that we would keep the part where the `ExternalProperties` are set and just delay the part where the `InProgressProperties` are cleared. The bind request was successful, so the state in the broker has changed, even if the Controller failed to inject the credentials.

Reply:
It probably doesn't matter too much what we do here in the short term. I think that long term, as part of reworking reconciliation, we will want the Controller to not send the bind request again once it is successful, even if the injection fails. At that point, the in-progress properties should be cleared and the external properties set as soon as the successful response is received, as they are currently in master.

Reply:
It seems like a logical error to me to have both `InProgressProperties` and `ExternalProperties` set. Either we are attempting to persist the properties in the broker, or the properties have been persisted.

The question then to me is what we should show the user. Should we give them the total truth of the situation that the properties have been persisted, even if it makes our API server validation less strong? Or should we tell a white lie that the properties are still in progress because it makes the system easier to reason about.

Reply:
As for the long-term discussion, I think we'll have to still send the bind request in some cases. For bindings that have `binding_retrievable` set, we could just do a GET, but not all bindings will be retrievable, and for those situations retrying the bind request will be the only way to get the credentials again should we have an error on our part persisting them locally.

Reply:
Ah, that makes sense with regards to having to repeat the bind request. I didn't think about the Controller needing to reacquire the information that is being injected.

Reply:
As for `InProgressProperties` and `ExternalProperties`, I don't consider it to be a white lie if we leave the in-progress properties while there is still a current operation, even if the broker has already accepted the bind. Is it a white lie that `CurrentOperation` is still set to Bind? The Controller is still making bind requests and those bind requests are still sending the properties in `InProgressProperties`.

Reply:
1. User creates a ServiceBinding with parameters sourced from a secret.
2. Controller sends a successful bind request to the broker.
3. Injecting the credentials fails for the ServiceBinding.
4. The user modifies the secret sourcing the parameters for the ServiceBinding.
5. Controller attempts more bind requests to which the broker responds with a 409 Conflict since the parameters sent are different than the parameters that were sent in the initial successful bind request.

In that scenario, there are bind requests occurring for an active bind operation. The parameters sent in the bind request do not match what is in `ExternalProperties` for the ServiceBinding. It seems to me that the `InProgressProperties` should match what is being sent to the broker.

Reply:
Hmm... Looking at the comment for `InProgressProperties` in `types.go`, it says:

```go
// InProgressProperties is the properties state of the
// ServiceBinding when a Bind is in progress. If the current
// operation is an Unbind, this will be nil.
InProgressProperties *ServiceBindingPropertiesState
```

Since the "Bind" operation as defined by the controller is still ongoing, I suppose according to that definition of the property it could still be set while we're attempting to inject the credentials.

And the `ExternalProperties` comment says:

```go
// ExternalProperties is the properties state of the
// ServiceBinding which the broker knows about.
ExternalProperties *ServiceBindingPropertiesState
```

Since we know the broker has persisted the properties, I suppose this would be set as well.

I would argue that the two *should* be mutually exclusive, and having two potentially-giant identical parameter payloads is just muddying up the binding object, but the definitions right now seem loose enough that there is overlap here.

I'd like to debate this further, but for the time being in order to fix this bug, according to our definitions it seems alright to have both in-progress and external properties set in this case. I'll make the change.